File: src/main/java/org/cubexmc/railway/train/TrainConsist.java
package org.cubexmc.railway.train;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.bukkit.entity.Minecart;
import org.bukkit.util.Vector;

public class TrainConsist {

    private final List<Minecart> cars = new ArrayList<>();

    public void addCar(Minecart car) {
        cars.add(car);
    }

    public List<Minecart> getCars() {
        return Collections.unmodifiableList(cars);
    }

    public Minecart getLeadCar() {
        return cars.isEmpty() ? null : cars.get(0);
    }

    public boolean contains(Minecart car) {
        return cars.contains(car);
    }

    public void setVelocity(Vector velocity) {
        for (Minecart car : cars) {
            if (car == null || car.isDead()) continue;
            car.setVelocity(velocity);
        }
    }

    public void zeroVelocity() {
        setVelocity(new Vector(0, 0, 0));
    }

    public void clear() {
        cars.clear();
    }
}

File: src/main/java/org/cubexmc/metro/train/TrainConsist.kt
package org.cubexmc.metro.train

import org.bukkit.entity.Minecart
import org.bukkit.util.Vector
import java.util.Collections

class TrainConsist {

    private val cars: MutableList<Minecart> = ArrayList()

    fun addCar(car: Minecart) {
        cars.add(car)
    }

    fun getCars(): List<Minecart> = Collections.unmodifiableList(cars)

    fun getLeadCar(): Minecart? = if (cars.isEmpty()) null else cars[0]

    fun contains(car: Minecart?): Boolean = cars.contains(car)

    fun setVelocity(velocity: Vector) {
        for (car in cars) {
            if (car.isDead) continue
            car.velocity = velocity
        }
    }

    fun zeroVelocity() {
        setVelocity(Vector(0, 0, 0))
    }

    fun clear() {
        cars.clear()
    }
}
